File: implementations/python/cli/ff3_stresstest.py
#!/usr/bin/env python3
"""FF3 Stress Test Tool - Randomized testing for stability."""

import argparse
import random
import secrets
import sys
import time
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from src.ff3 import (
    digits,
    hex_lower,
    base36_lower,
    base62,
    ALPHA_DIGITS,
    ALPHA_HEX_LOWER,
    ALPHA_BASE36_LOW,
    ALPHA_BASE62,
)

DEFAULT_ITERATIONS = 1000
DEFAULT_MIN_LEN = 6
DEFAULT_MAX_LEN = 20
DEFAULT_ALPHABETS = ["digits", "hex-lower", "base36-lower", "base62"]

ALPHABET_MAP = {
    "digits": (ALPHA_DIGITS, digits),
    "hex-lower": (ALPHA_HEX_LOWER, hex_lower),
    "base36-lower": (ALPHA_BASE36_LOW, base36_lower),
    "base62": (ALPHA_BASE62, base62),
}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="FF3 Stress Test Tool",
        formatter_class=argparse.RawTextHelpFormatter,
    )
    parser.add_argument(
        "iterations",
        nargs="?",
        type=int,
        default=DEFAULT_ITERATIONS,
        help="Number of iterations per alphabet (default: 1000)",
    )
    parser.add_argument(
        "--alphabets",
        type=str,
        default="",
        help="Comma-separated list (default: digits,hex-lower,base36-lower,base62)",
    )
    parser.add_argument(
        "--min-length",
        type=int,
        default=DEFAULT_MIN_LEN,
        help="Minimum plaintext length (default: 6)",
    )
    parser.add_argument(
        "--max-length",
        type=int,
        default=DEFAULT_MAX_LEN,
        help="Maximum plaintext length (default: 20)",
    )
    parser.add_argument(
        "--quick",
        action="store_true",
        help="Run 100 iterations (fast test)",
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=None,
        help="Random seed for reproducibility",
    )
    return parser.parse_args()


def validate_options(ns: argparse.Namespace) -> tuple[int, list[str], int, int, int]:
    iterations = ns.iterations
    if ns.quick:
        iterations = 100
    if iterations <= 0:
        raise ValueError("iterations must be greater than 0")

    min_len = ns.min_length
    max_len = ns.max_length
    if min_len < 2:
        raise ValueError("--min-length must be at least 2")
    if max_len < min_len:
        raise ValueError("--max-length must be greater than or equal to --min-length")

    if ns.alphabets.strip() == "":
        alphabets = DEFAULT_ALPHABETS
    else:
        alphabets = [name.strip() for name in ns.alphabets.split(",") if name.strip()]
        if not alphabets:
            raise ValueError("--alphabets list cannot be empty")

    unknown = [name for name in alphabets if name not in ALPHABET_MAP]
    if unknown:
        raise ValueError(f"unknown alphabet(s): {', '.join(unknown)}")

    seed = ns.seed
    return iterations, alphabets, min_len, max_len, seed


def generate_plaintext(charset: str, length: int, rng: random.Random) -> str:
    return "".join(rng.choice(charset) for _ in range(length))


def print_failure(key: bytes, tweak: bytes, plaintext: str, ciphertext: str, decrypted: str | None, detail: str | None) -> None:
    print("  Round-trip failed:")
    print(f"    Key: {key.hex()}")
    print(f"    Tweak: {tweak.hex()}")
    print(f"    Plaintext: \"{plaintext}\"")
    if ciphertext:
        print(f"    Ciphertext: \"{ciphertext}\"")
    if decrypted:
        print(f"    Decrypted: \"{decrypted}\"")
    if detail:
        print(f"    Detail: {detail}")


def stress_test_alphabet(
    name: str,
    charset: str,
    factory,
    iterations: int,
    min_len: int,
    max_len: int,
    rng: random.Random,
    byte_source,
) -> tuple[int, int]:
    print(f"Testing {name}...")
    print(f"  Alphabet: {charset} (radix {len(charset)})")

    passed = 0
    failed = 0
    interval = max(1, iterations // 10)

    for i in range(iterations):
        key = byte_source(16)
        tweak = byte_source(8)
        length = rng.randint(min_len, max_len)
        plaintext = generate_plaintext(charset, length, rng)

        try:
            cipher = factory(key, tweak)
        except Exception as exc:  # pylint: disable=broad-except
            failed += 1
            print_failure(key, tweak, plaintext, "", None, f"cipher initialization failed: {exc}")
            continue

        try:
            ciphertext = cipher.encrypt(plaintext)
            decrypted = cipher.decrypt(ciphertext)
        except Exception as exc:  # pylint: disable=broad-except
            failed += 1
            print_failure(key, tweak, plaintext, "", None, f"encryption error: {exc}")
            continue

        if decrypted != plaintext:
            failed += 1
            print_failure(key, tweak, plaintext, ciphertext, decrypted, "round-trip mismatch")
        else:
            passed += 1

        if (i + 1) % interval == 0 or (i + 1) == iterations:
            percent = (i + 1) * 100 // iterations
            print(f"  Progress: {i + 1}/{iterations} ({percent}%)")

    print(f"  Passed: {passed}/{passed + failed}")
    print(f"  Failed: {failed}/{passed + failed}\n")
    return passed, failed


def main() -> int:
    try:
        args = parse_args()
        iterations, alphabets, min_len, max_len, seed = validate_options(args)
    except ValueError as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    rng = random.Random(seed) if seed is not None else random.Random()
    if seed is not None:
        def byte_source(size: int) -> bytes:
            return bytes(rng.randrange(0, 256) for _ in range(size))
    else:
        def byte_source(size: int) -> bytes:
            return secrets.token_bytes(size)

    print("FF3 Stress Test v1.0")
    print("====================\n")
    print("Warning: FF3 was withdrawn by NIST; run for education and research only.\n")
    print("Test configuration")
    print(f"  Iterations per alphabet: {iterations}")
    print("  Random key/tweak generation: enabled")
    print(f"  String length range: {min_len}-{max_len} characters")
    print(f"  Alphabets: {', '.join(alphabets)}\n")

    total_tests = 0
    total_failures = 0
    start = time.time()

    for name in alphabets:
        charset, factory = ALPHABET_MAP[name]
        passed, failed = stress_test_alphabet(name, charset, factory, iterations, min_len, max_len, rng, byte_source)
        total_tests += passed + failed
        total_failures += failed

    duration_ms = int((time.time() - start) * 1000)

    print("Summary")
    print(f"  Total tests: {total_tests}")
    print(f"  Failures: {total_failures}")
    if duration_ms > 0:
        throughput = (total_tests * 1000) / duration_ms
        print(f"  Duration: {duration_ms} ms")
        print(f"  Throughput: {throughput:.2f} tests/sec")
    if total_failures == 0:
        print("  Result: all stress tests passed")
        return 0
    print("  Result: failures detected")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: implementations/erlang/apps/ff3_stresstest/src/ff3_stresstest.erl
%% FF3 Stress Test Tool - Erlang Implementation

-module(ff3_stresstest).
-export([main/1]).

-define(DEFAULT_ITERATIONS, 1000).
-define(DEFAULT_MIN_LEN, 6).
-define(DEFAULT_MAX_LEN, 20).
-define(DEFAULT_ALPHABETS, ["digits", "hex-lower", "base36-lower", "base62"]).

-record(opts, {
    iterations = ?DEFAULT_ITERATIONS :: pos_integer(),
    alphabets = ?DEFAULT_ALPHABETS :: [string()],
    min_len = ?DEFAULT_MIN_LEN :: pos_integer(),
    max_len = ?DEFAULT_MAX_LEN :: pos_integer(),
    quick = false :: boolean(),
    seed = undefined :: undefined | non_neg_integer()
}).

main(Args) ->
    case parse_args(Args, #opts{}) of
        {error, Reason} ->
            io:format("Error: ~s~n", [Reason]),
            erlang:halt(1);
        {ok, Opts0} ->
            Opts = maybe_apply_quick(Opts0),
            case validate_opts(Opts) of
                ok -> run(Opts);
                {error, Reason2} ->
                    io:format("Error: ~s~n", [Reason2]),
                    erlang:halt(1)
            end
    end.

maybe_apply_quick(#opts{quick = true} = Opts) -> Opts#opts{iterations = 100};
maybe_apply_quick(Opts) -> Opts.

validate_opts(#opts{iterations = Iter}) when Iter =< 0 ->
    {error, "iterations must be greater than 0"};
validate_opts(#opts{min_len = Min}) when Min < 2 ->
    {error, "--min-length must be at least 2"};
validate_opts(#opts{min_len = Min, max_len = Max}) when Max < Min ->
    {error, "--max-length must be greater than or equal to --min-length"};
validate_opts(_) -> ok.

parse_args([], Opts) -> {ok, Opts};
parse_args(["--help" | _], _Opts) ->
    print_usage(),
    erlang:halt(0);
parse_args(["-h" | Rest], Opts) -> parse_args(["--help" | Rest], Opts);
parse_args(["--alphabets", Value | Rest], Opts) ->
    Alphs = [string:trim(A) || A <- string:tokens(Value, ","), A =/= ""],
    case Alphs of
        [] -> {error, "--alphabets list cannot be empty"};
        _ -> parse_args(Rest, Opts#opts{alphabets = Alphs})
    end;
parse_args(["--min-length", Value | Rest], Opts) ->
    case string:to_integer(Value) of
        {error, _} -> {error, "invalid value for --min-length"};
        {Min, _} -> parse_args(Rest, Opts#opts{min_len = Min})
    end;
parse_args(["--max-length", Value | Rest], Opts) ->
    case string:to_integer(Value) of
        {error, _} -> {error, "invalid value for --max-length"};
        {Max, _} -> parse_args(Rest, Opts#opts{max_len = Max})
    end;
parse_args(["--quick" | Rest], Opts) ->
    parse_args(Rest, Opts#opts{quick = true});
parse_args(["--seed", Value | Rest], Opts) ->
    case string:to_integer(Value) of
        {error, _} -> {error, "invalid value for --seed"};
        {Seed, _} when Seed >= 0 -> parse_args(Rest, Opts#opts{seed = Seed});
        _ -> {error, "--seed must be non-negative"}
    end;
parse_args([Value | Rest], Opts) ->
    case string:to_integer(Value) of
        {error, _} ->
            {error, lists:flatten(io_lib:format("unknown option: ~s", [Value]))};
        {Iterations, _} ->
            parse_args(Rest, Opts#opts{iterations = Iterations})
    end.

run(#opts{seed = Seed} = Opts) ->
    case Seed of
        undefined -> rand:seed(exsss, now_seed());
        S -> rand:seed(exsss, {S, S, S})
    end,

    io:format("FF3 Stress Test v1.0~n"),
    io:format("====================~n~n"),
    io:format("Warning: FF3 was withdrawn by NIST; run for education and research only.~n~n"),
    io:format("Test configuration~n"),
    io:format("  Iterations per alphabet: ~p~n", [Opts#opts.iterations]),
    io:format("  Random key/tweak generation: enabled~n"),
    io:format("  String length range: ~p-~p characters~n", [Opts#opts.min_len, Opts#opts.max_len]),
    io:format("  Alphabets: ~s~n~n", [string:join(Opts#opts.alphabets, ", ")]),

    AlphabetInfos = [alphabet_info(Name) || Name <- Opts#opts.alphabets],
    Start = erlang:monotonic_time(millisecond),

    {TotalTests, TotalFailures} = lists:foldl(
        fun(Info, {AccTests, AccFailures}) ->
            {Tests, Failures} = stress_alphabet(Info, Opts),
            {AccTests + Tests, AccFailures + Failures}
        end,
        {0, 0},
        AlphabetInfos
    ),

    Duration = erlang:monotonic_time(millisecond) - Start,

    io:format("Summary~n"),
    io:format("  Total tests: ~p~n", [TotalTests]),
    io:format("  Failures: ~p~n", [TotalFailures]),
    io:format("  Duration: ~p ms~n", [Duration]),
    case Duration > 0 of
        true ->
            Throughput = (TotalTests * 1000.0) / Duration,
            io:format("  Throughput: ~.2f tests/sec~n", [Throughput]);
        false -> ok
    end,
    case TotalFailures of
        0 ->
            io:format("  Result: all stress tests passed~n"),
            erlang:halt(0);
        _ ->
            io:format("  Result: failures detected~n"),
            erlang:halt(1)
    end.

alphabet_info("digits") -> {"digits", ff3_alphabets:digits_alphabet(), fun ff3_api:create_digits_cipher/2};
alphabet_info("hex-lower") -> {"hex-lower", ff3_alphabets:hex_lower_alphabet(), fun ff3_api:create_hex_lower_cipher/2};
alphabet_info("base36-lower") -> {"base36-lower", ff3_alphabets:base36_lower_alphabet(), fun ff3_api:create_base36_lower_cipher/2};
alphabet_info("base62") -> {"base62", ff3_alphabets:base62_alphabet(), fun ff3_api:create_base62_cipher/2};
alphabet_info(Name) -> throw({unknown_alphabet, Name}).

stress_alphabet({Name, Alphabet, Factory}, Opts) ->
    io:format("Testing ~s...~n", [Name]),
    io:format("  Alphabet: ~s (radix ~p)~n", [Alphabet, length(Alphabet)]),

    Iterations = Opts#opts.iterations,
    Interval = max(1, Iterations div 10),

    {Passed, Failed} = stress_loop(Iterations, Interval, 0, 0, Alphabet, Factory, Opts),

    io:format("  Passed: ~p/~p~n", [Passed, Iterations]),
    io:format("  Failed: ~p/~p~n~n", [Failed, Iterations]),
    {Iterations, Failed}.

stress_loop(0, _Interval, Passed, Failed, _Alphabet, _Factory, _Opts) -> {Passed, Failed};
stress_loop(Remaining, Interval, Passed, Failed, Alphabet, Factory, Opts) ->
    Counter = Opts#opts.iterations - Remaining + 1,
    maybe_print_progress(Counter, Interval, Opts#opts.iterations),

    Key = random_bytes(16),
    Tweak = random_bytes(8),
    Len = Opts#opts.min_len + rand:uniform(Opts#opts.max_len - Opts#opts.min_len + 1) - 1,
    Plain = generate_plaintext(Alphabet, Len),

    case run_iteration(Factory, Plain, Key, Tweak) of
        ok -> stress_loop(Remaining - 1, Interval, Passed + 1, Failed, Alphabet, Factory, Opts);
        {error, Detail} ->
            print_failure(Key, Tweak, Plain, Detail),
            stress_loop(Remaining - 1, Interval, Passed, Failed + 1, Alphabet, Factory, Opts)
    end.

maybe_print_progress(_, _, 0) -> ok;
maybe_print_progress(Counter, Interval, Total) when Counter rem Interval =:= 0; Counter =:= Total ->
    Percent = (Counter * 100) div Total,
    io:format("  Progress: ~p/~p (~p%)~n", [Counter, Total, Percent]);
maybe_print_progress(_, _, _) -> ok.

run_iteration(Factory, Plain, Key, Tweak) ->
    case Factory(Key, Tweak) of
        {ok, Cipher} ->
            case ff3_api:encrypt_string(Cipher, Plain) of
                {ok, Ciphertext} ->
                    case ff3_api:decrypt_string(Cipher, Ciphertext) of
                        {ok, Plain} -> ok;
                        {ok, Decrypted} -> {error, {ciphertext, Ciphertext, Decrypted}};
                        {error, Reason} -> {error, Reason}
                    end;
                {error, Reason} -> {error, Reason}
            end;
        {error, Reason} -> {error, Reason}
    end.

print_failure(Key, Tweak, Plain, {ciphertext, Ciphertext, Decrypted}) ->
    io:format("  Round-trip failed:\n"),
    io:format("    Key: ~s\n", [hex(Key)]),
    io:format("    Tweak: ~s\n", [hex(Tweak)]),
    io:format("    Plaintext: \"~s\"\n", [Plain]),
    io:format("    Ciphertext: \"~s\"\n", [Ciphertext]),
    io:format("    Decrypted: \"~s\"\n", [Decrypted]);
print_failure(Key, Tweak, Plain, Detail) ->
    io:format("  Round-trip failed:\n"),
    io:format("    Key: ~s\n", [hex(Key)]),
    io:format("    Tweak: ~s\n", [hex(Tweak)]),
    io:format("    Plaintext: \"~s\"\n", [Plain]),
    io:format("    Detail: ~p\n", [Detail]).

random_bytes(N) -> crypto:strong_rand_bytes(N).

hex(Bytes) -> string:to_lower(binary_to_list(binary:encode_hex(Bytes))).

generate_plaintext(Alphabet, Len) ->
    [lists:nth(rand:uniform(length(Alphabet)), Alphabet) || _ <- lists:seq(1, Len)].

now_seed() ->
    {Mega, Sec, Micro} = os:timestamp(),
    {Mega rem 30000 + 1, Sec rem 30000 + 1, Micro rem 30000 + 1}.

max(A, B) when A >= B -> A;
max(_, B) -> B.

print_usage() ->
    io:format("FF3 Stress Test Tool~n~n"),
    io:format("Usage: ff3_stresstest [OPTIONS] [ITERATIONS]~n~n"),
    io:format("Options:~n"),
    io:format("  --alphabets LIST      Comma-separated list (default: digits,hex-lower,base36-lower,base62)~n"),
    io:format("  --min-length N        Minimum plaintext length (default: 6)~n"),
    io:format("  --max-length N        Maximum plaintext length (default: 20)~n"),
    io:format("  --quick               Run 100 iterations (fast test)~n"),
    io:format("  --seed N              Random seed for reproducibility~n").
